File: app/src/main/java/com/jet/mvvmtest/MainActivity.java
package com.jet.mvvmtest;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.BindingAdapter;
import androidx.databinding.DataBindingUtil;

import android.app.Activity;
import android.content.Context;
import android.content.ContextWrapper;
import android.os.Bundle;
import android.os.Handler;
import android.text.TextUtils;
import android.view.View;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.jet.mvvmtest.databinding.ActivityMainBinding;

public class MainActivity extends AppCompatActivity {

    public ActivityMainBinding getActivityMainBinding() {
        return activityMainBinding;
    }

    private ActivityMainBinding activityMainBinding;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        activityMainBinding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        activityMainBinding.setViewModel(new LoginViewModel());
        activityMainBinding.executePendingBindings();
    }

    @BindingAdapter({"clicked"})
    public static void checkClicked(View view, boolean clicked) {
        Activity activity = getActivity(view.getContext());
        if(activity == null){
            return;
        }
        MainActivity mainActivity = (MainActivity)activity;
        if(clicked) {
            EditText textViewUsername = mainActivity.getActivityMainBinding().username;
            if (TextUtils.isEmpty(textViewUsername.getText())) {
                textViewUsername.setError(view.getContext().getString(R.string.username_error));
                return;
            }
            EditText textViewPassword = mainActivity.getActivityMainBinding().password;
            if (TextUtils.isEmpty(textViewPassword.getText())) {
                textViewPassword.setError(view.getContext().getString(R.string.password_error));
                return;
            }

            mainActivity.getActivityMainBinding().progress.setVisibility(View.VISIBLE);
            new Handler().postDelayed(() -> {
                Toast.makeText(view.getContext(), R.string.log_in_succeed, Toast.LENGTH_SHORT).show();
                mainActivity.getActivityMainBinding().progress.setVisibility(View.GONE);
            }, 2000);

        }
    }

    public static Activity getActivity(Context context)
    {
        if (context == null)
        {
            return null;
        }
        else if (context instanceof ContextWrapper)
        {
            if (context instanceof Activity)
            {
                return (Activity) context;
            }
            else
            {
                return getActivity(((ContextWrapper) context).getBaseContext());
            }
        }

        return null;
    }

}

File: app/src/main/java/com/jet/mvvmtest/MainActivity.kt
package com.jet.mvvmtest

import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.BindingAdapter
import androidx.databinding.DataBindingUtil

import android.app.Activity
import android.content.Context
import android.content.ContextWrapper
import android.os.Bundle
import android.os.Handler
import android.text.TextUtils
import android.view.View
import android.widget.EditText
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast

import com.jet.mvvmtest.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {

    var activityMainBinding: ActivityMainBinding? = null
        private set

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        activityMainBinding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        activityMainBinding!!.viewModel = LoginViewModel()
        activityMainBinding!!.executePendingBindings()
    }

    companion object {

        @BindingAdapter("clicked")
        fun checkClicked(view: View, clicked: Boolean) {
            val activity = getActivity(view.context) ?: return
            val mainActivity = activity as MainActivity
            if (clicked) {
                val textViewUsername = mainActivity.activityMainBinding!!.username
                if (TextUtils.isEmpty(textViewUsername.text)) {
                    textViewUsername.error = view.context.getString(R.string.username_error)
                    return
                }
                val textViewPassword = mainActivity.activityMainBinding!!.password
                if (TextUtils.isEmpty(textViewPassword.text)) {
                    textViewPassword.error = view.context.getString(R.string.password_error)
                    return
                }

                mainActivity.activityMainBinding!!.progress.visibility = View.VISIBLE
                Handler().postDelayed({
                    Toast.makeText(view.context, R.string.log_in_succeed, Toast.LENGTH_SHORT).show()
                    mainActivity.activityMainBinding!!.progress.visibility = View.GONE
                }, 2000)

            }
        }

        fun getActivity(context: Context?): Activity? {
            if (context == null) {
                return null
            } else if (context is ContextWrapper) {
                return if (context is Activity) {
                    context
                } else {
                    getActivity(context.baseContext)
                }
            }

            return null
        }
    }

}
